baseapp/abci.go

```diff
@@ -1028,9 +1028,9 @@ func handleQueryApp(app *BaseApp, path []string, req *abci.RequestQuery) *abci.R
 			txBytes := req.Data
 
 			// Simulate enters runTx which requires us to give up the lock now.
-			app.mtx.Unlock()
+			app.mtx.RUnlock()
 			gInfo, res, err := app.Simulate(txBytes)
```

Comment:
I don't think golang RWMutex is reentrant (see below comment), but if it were: When recursively calling Query in grpc query case, does it always go back to grpc in the inner call as well? Or in other words, what prevents the inner call from reaching this handleQueryApp case (and deadlocking due to trying to acquire lock)?

Moot point if you re-add unlocking and relocking for read lock below.

Reply:
Yeah, I'm re-adding the RUnlock/Lock back.